File: dei_frontend/src/app/core/services/user.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { API } from '../constants/api-paths';
import { apiUrl, pageParams, toParams } from './http-util';
import { Page, PageQuery } from '../models/common.model';
import { Role } from '../models/enums';
import {
  AdminCreateUserRequest,
  AdminUpdateUserRequest,
  ScopeValueOption,
  UpdateProfileRequest,
  UserResponse,
} from '../models/iam.model';

@Injectable({ providedIn: 'root' })
export class UserService {
  private readonly http = inject(HttpClient);

  list(page: PageQuery, role?: Role | null): Observable<Page<UserResponse>> {
    return this.http.get<Page<UserResponse>>(apiUrl(API.users.base), {
      params: pageParams(page, { role: role ?? undefined }),
    });
  }

  listByRole(role: Role, size = 100): Observable<Page<UserResponse>> {
    return this.http.get<Page<UserResponse>>(apiUrl(API.users.base), {
      params: toParams({ role, page: 0, size }),
    });
  }

  me(): Observable<UserResponse> {
    return this.http.get<UserResponse>(apiUrl(API.users.me));
  }

  scopeValues(scope: 'DEPARTMENT' | 'GRADE'): Observable<ScopeValueOption[]> {
    return this.http.get<ScopeValueOption[]>(apiUrl(API.users.scopeValues), {
      params: toParams({ scope }),
    });
  }

  getById(id: number): Observable<UserResponse> {
    return this.http.get<UserResponse>(apiUrl(API.users.byId(id)));
  }

  create(body: AdminCreateUserRequest): Observable<UserResponse> {
    return this.http.post<UserResponse>(apiUrl(API.users.base), body);
  }

  updateProfile(body: UpdateProfileRequest): Observable<UserResponse> {
    return this.http.put<UserResponse>(apiUrl(API.users.me), body);
  }

  adminUpdate(id: number, body: AdminUpdateUserRequest): Observable<UserResponse> {
    return this.http.put<UserResponse>(apiUrl(API.users.byId(id)), body);
  }

  deactivate(id: number): Observable<void> {
    return this.http.delete<void>(apiUrl(API.users.byId(id)));
  }
}

File: dei_backend/dei-connect/src/main/java/com/deiconnect/iam/service/UserService.java
package com.deiconnect.iam.service;

import com.deiconnect.common.enums.Role;
import com.deiconnect.iam.dto.AdminCreateUserRequest;
import com.deiconnect.iam.dto.AdminUpdateUserRequest;
import com.deiconnect.iam.dto.ScopeValueOption;
import com.deiconnect.iam.dto.UpdateProfileRequest;
import com.deiconnect.iam.dto.UserResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface UserService {

    UserResponse adminCreate(AdminCreateUserRequest request);

    UserResponse getCurrentProfile();

    UserResponse getById(Long id);

    UserResponse getByIdInternal(Long id);

    List<UserResponse> getByIdsInternal(List<Long> ids);

    List<UserResponse> getEmployeesByManagerInternal(Long managerId);

    Page<UserResponse> list(Role role, Pageable pageable);

    UserResponse updateCurrentProfile(UpdateProfileRequest request);

    UserResponse adminUpdate(Long id, AdminUpdateUserRequest request);

    void deactivate(Long id);

    List<ScopeValueOption> getScopeValues(String scope);
}
